### PR TITLE
Canonicalize single bit comparisons

### PR DESCRIPTION
**Pull request by prithayan**
Canonicalize the following comparisons, (https://github.com/llvm/circt/issues/634)

   Case1:   x == 0 -> x ^ 1
   Case2:   x == 1 -> x
   Case3:   x != 0 -> x
   Case4:   x != 1 -> x ^ 1

### CODE REVIEW

File: lib/Dialect/Comb/CombFolds.cpp [hunk 1]
```diff
@@ -952,6 +952,34 @@ struct ICmpCanonicalizeConstant final : public OpRewritePattern<ICmpOp> {
...
         return replaceWith(ICmpPredicate::ugt, op.lhs(), getConstant(rhs - 1));
+      case ICmpPredicate::eq:
+        if (rhs.getBitWidth() != 1)
+          break;
+        if (rhs.isNullValue()) {
+          // x == 0 -> x ^ 1
+          rewriter.replaceOpWithNewOp<XorOp>(op, op.lhs(),
+                                             getConstant(APInt(1, 1)));
+          return success();
+        } else if (rhs.isAllOnesValue()) {
```
Comment (seldridge) [inline lib/Dialect/Comb/CombFolds.cpp:963]: Nit: no need for the `else` here. Just an `if` is fine.

File: lib/Dialect/Comb/CombFolds.cpp [hunk 2]
```diff
@@ -952,6 +952,34 @@ struct ICmpCanonicalizeConstant final : public OpRewritePattern<ICmpOp> {
...
+        }
+        break;
+      case ICmpPredicate::ne:
+        if (rhs.getBitWidth() != 1)
+          break;
+        if (rhs.isNullValue()) {
+          // x != 0 -> x
+          rewriter.replaceOp(op, op.lhs());
+          return success();
+        } else if (rhs.isAllOnesValue()) {
```
Comment (seldridge) [inline lib/Dialect/Comb/CombFolds.cpp:976]: Nit: No need for `else`.

File: test/Dialect/RTL/canonicalization.mlir [hunk 3]
```diff
@@ -766,6 +766,43 @@ rtl.module @shrs_fold2() -> (i12) {
...
+ // CHECK-LABEL: rtl.module @icmp_fold_1bit_neq2(%arg: i1) -> (i1) {
+// CHECK-NEXT:    %true = comb.constant true
+// CHECK-NEXT:    %0 = comb.xor %arg, %true : i1
+// CHECK-NEXT:    rtl.output %0 : i1
+
+rtl.module @icmp_fold_1bit_neq2(%arg: i1) -> (i1) {
+  %c1 = comb.constant 1 : i1
+  %0 = comb.icmp ne  %c1, %arg  : i1
+  rtl.output %0 : i1
+}
```
Comment (seldridge) [inline test/Dialect/RTL/canonicalization.mlir:804]: Nit: I'd make the comments here look like the comments above:

1. No space between the comment and the test it's associated with
1. One space after each test
1. Keep everything flush left (remove any of the extra spaces included on the left)

**Review by seldridge — APPROVED** [review]
Excellent. 

Thanks for fleshing out these canonicalizations. :+1:

File: test/Dialect/RTL/canonicalization.mlir (hunk 3 above)
Comment (darthscsi) [reply test/Dialect/RTL/canonicalization.mlir:804]: Other nit, fold all the bodies into one module.  No need to have a separate module per condition.  I realize this is somewhat ad-hoc, but for simple checks of the same set of folds, it seems nicer to not make lots of extra text.

**Review by darthscsi — APPROVED** [review]
LGTM, pending addressing @seldridge 's comments.

File: test/Dialect/RTL/canonicalization.mlir (hunk 3 above)
Comment (prithayan) [reply test/Dialect/RTL/canonicalization.mlir]: Updated the test cases into a single module

**Review by lattner — COMMENTED** [review]
nice!

**Pull request merged**